Ignore wrapped exception `gate_closed_exception` and `rpc::closed_error` when node shuts down.

bitpathfinder opened the pull request:
Normally, when a node is shutting down, `gate_closed_exception` and `rpc::closed_error`
in `send_to_live_endpoints` should be ignored. However, if these exceptions are wrapped
in a `nested_exception`, an error message is printed, causing tests to fail.
    
This commit adds handling for nested exceptions in this case to prevent unnecessary
error messages.
    
Fixes scylladb/scylladb#23325

mykaul commented:
> Backport: not needed, since this is relatively rare issue.

I'd backport it to 2025.1, at least. We'll see it forever in CI of that LTS release.

mykaul commented:
Does it fix https://github.com/scylladb/scylladb/issues/23305 as well?

mykaul commented:
https://github.com/scylladb/scylladb/issues/21815 ?

bitpathfinder commented:
> Does it fix #23305 as well?

Yes, it looks like the same issue to me.

bitpathfinder commented:
> #21815 ?

Yep, that one as well.

avikivity commented:
I'm not thrilled about our exception handling strategy. gate_closed_exception is very generic and says nothing about whether it was important or not.

That said, the patch doesn't create the mess.

piodul reviewed (changes requested):
I don't think the fix does what it is supposed to do.

First, `try_catch<T>` simulates the `catch (const T&)` semantics. In particular, if the exception is a class _derived_ from `T`, then `try_catch<T>` will catch it, too.

`utils::internal::nested_exception<T>` inherits from `T`, so it is supposed to be caught by `try_catch<T>`, too. Therefore the new `try_catch_nested<T>` should already do the same as the existing `try_catch<T>`.

Looking at the logs from the description of issue https://github.com/scylladb/scylladb/issues/23325:

```
storage_proxy - exception during mutation write to 3052755d-2317-42e2-a434-c970f2df2e53: utils::internal::nested_exception<std::runtime_error> (Could not write mutation ks:cf (pk{00056b31303331}) to commitlog): seastar::gate_closed_exception (gate closed)
```

It's a nested_exception for __std::runtime_error__, and not seastar::gate_closed_exception. I think that the problem might come from inside the `database::apply_with_commitlog` function:

```c++
future<> database::apply_with_commitlog(column_family& cf, const mutation& m, db::timeout_clock::time_point timeout) {
    db::rp_handle h;
    if (cf.commitlog() != nullptr && cf.durable_writes()) {
        auto fm = freeze(m);
        std::exception_ptr ex;
        try {
            commitlog_entry_writer cew(m.schema(), fm, db::commitlog::force_sync::no);
            auto f_h = co_await coroutine::as_future(cf.commitlog()->add_entry(m.schema()->id(), cew, timeout));
            if (!f_h.failed()) {
                h = f_h.get();
            } else {
                ex = f_h.get_exception();
            }
        } catch (...) {
            ex = std::current_exception();
        }
        if (ex) {
            if (try_catch<timed_out_error>(ex)) {
                ex = wrap_commitlog_add_error<wrapped_timed_out_error>(cf.schema(), fm, std::move(ex));
            } else {
                ex = wrap_commitlog_add_error<>(cf.schema(), fm, std::move(ex));
            }
            co_await coroutine::exception(std::move(ex));
        }
    }
    try {
        co_await apply_in_memory(m, cf, std::move(h), timeout);
    } catch (mutation_reordered_with_truncate_exception&) {
        // This mutation raced with a truncate, so we can just drop it.
        dblog.debug("replay_position reordering detected");
    }
}
```

We only explicitly handle `timed_out_error` there. We should handle `seastar::gate_closed_exception` there as well.

piodul commented:
> We only explicitly handle `timed_out_error` there. We should handle `seastar::gate_closed_exception` there as well.

We actually have a similar piece of logic in `database::do_apply` as well. Perhaps it would be a good idea to unify the error handling logic in both functions.

bitpathfinder commented:
```C++
                ex = wrap_commitlog_add_error<>(cf.schema(), fm, std::move(ex));
```
@piodul Why don't we preserve the original exception there? Why hide it under "runtime_error"?

piodul commented:
> ```c++
>                 ex = wrap_commitlog_add_error<>(cf.schema(), fm, std::move(ex));
> ```
> 
> @piodul Why don't we preserve the original exception there? Why hide it under "runtime_error"?

The way that `make_nested_exception_ptr<T>` works (which simulates [`std::throw_with_nested`](https://en.cppreference.com/w/cpp/error/throw_with_nested)) is that it returns an exception which is derived from `T` and from `std::nested_exception`. Because of that, you need to know the type of the exception in compile time, so unless you handle each case explicitly (like we do for `timed_out_error`/`wrapped_timed_out_error`)  you will need a fallback like this.

I am not the original author, this was introduced back in https://github.com/scylladb/scylladb/pull/9412/commits/a6202ae079b44da44f402f50001bd2df1cb49673. The intent seems to be that we want to some attach context to the timeout, i.e. where the exception comes from. Unfortunately, using `std::runtime_error` unconditionally breaks the exception handling code, so it was attempted to be mitigated in https://github.com/scylladb/scylladb/commit/97bb1be6f7bae22fa769416152db71522cdb9970 with the use of `wrapped_timed_out_error`.

I'm not sure if I have a good suggestion on how to clean up the exception handling logic here. I suppose that it's rather rare that commitlog starts throwing new types of exceptions, so perhaps just adding an explicit `wrap_commitlog_add_error` here should be fine.

cc: @elcallio

elcallio commented on utils/exceptions.hh:
```diff
@@ -149,6 +149,16 @@ inline T* try_catch(std::exception_ptr& eptr) noexcept {
+    if (res) {
+        return res;
+    }
+    return try_catch<utils::internal::nested_exception<TException>>(eptr);
```
Would it not be prudent to also handle std::nested_exceptions here? I.e. if second try-catch fails, check for std::nested (std::rethrow_if_nested etc)?

bitpathfinder replied:
As @piodul noticed in [this comment](https://github.com/scylladb/scylladb/pull/23336#pullrequestreview-2694360214) this patch does not fix the issue (and does not do what intended), so I'm going to rewrite it.

elcallio commented:
`wrap_commitlog_add_error` only exists because of the need to know the context of the failure, indeed, because the lower layers have no idea of _what_ we are committing, and from where.

bitpathfinder commented:
> > We only explicitly handle `timed_out_error` there. We should handle `seastar::gate_closed_exception` there as well.
> 
> We actually have a similar piece of logic in `database::do_apply` as well. Perhaps it would be a good idea to unify the error handling logic in both functions.

looking at `do_apply`, why does it increment `_stats->total_writes_timedout` counter, in the copy-pasted `apply_with_commitlog` part, whereas `apply_with_commitlog` itself does not increment it?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16105/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4055/)
:x: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15208/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7173/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16105/)
#### Failed Tests (2/43122):
- [test_replace_node_same_ip_take_write[rbo_disabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16105/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_gossip_topology_changes___test_replace_node_same_ip_take_write_rbo_disabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_node_same_ip_take_write[rbo_disabled])
- [test_random_failures[stop_after_bootstrapping_initial_raft_configuration-add_new_node].dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16105/testReport/junit/cluster.random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_bootstrapping_initial_raft_configuration_add_new_node__dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_bootstrapping_initial_raft_configuration-add_new_node].dev.1)
#### Build Details:
- Duration: 6 hr 32 min
- Builder: i-099f1e0a61ef2c378 (m5d.16xlarge)

piodul commented:
> > > We only explicitly handle `timed_out_error` there. We should handle `seastar::gate_closed_exception` there as well.
> > 
> > 
> > We actually have a similar piece of logic in `database::do_apply` as well. Perhaps it would be a good idea to unify the error handling logic in both functions.
> 
> looking at `do_apply`, why does it increment `_stats->total_writes_timedout` counter, in the copy-pasted `apply_with_commitlog` part, whereas `apply_with_commitlog` itself does not increment it?

`apply_with_commitlog` is called from `do_apply_counter_update`, which is called within the context of `update_write_metrics`, which in turn takes care to update the `total_writes_timedout` metric. In that case, unifying the two places would require a larger refactor, so perhaps it is not worth doing in this PR.

bitpathfinder commented:
New Version: 

- Fixed implementation of `try_catch_nested`, added tests.
- Unified exceptions wrapping in `database::do_apply` and `database::apply_with_commitlog`.

elcallio reviewed (approved):
I still would like to know _why_ we get gate closed etc here. We should really ensure shutdown order ensures it does not happen. But outside this PR.

bitpathfinder commented:
> I still would like to know _why_ we get gate closed etc here. We should really ensure shutdown order ensures it does not happen. But outside this PR.

Agree, honestly I don't like the solution, since the root cause is the services shutdown order.

mykaul commented:
@emdroid , @bitpathfinder - what's the next step here? CI is utterly broken.

bitpathfinder commented:
> @emdroid , @bitpathfinder - what's the next step here? CI is utterly broken.

This PR is waiting for review and merge.
@piodul could you re-review please?

emdroid commented on utils/exceptions.hh:
```diff
@@ -149,6 +151,29 @@ inline T* try_catch(std::exception_ptr& eptr) noexcept {
+/// The same as try_catch, but also unwraps if the exception was nested.
+template <typename TException>
+inline TException* try_catch_nested(std::exception_ptr& eptr) noexcept {
+    /// Check if we got the exception we were looking for in the upper level.
```
should have just "//" comments inside the function ... I see that this file uses the Doxygen style "///" comments, but only for the function-level docs

emdroid commented on test/boost/exceptions_test.inc.cc:
```diff
@@ -136,6 +136,73 @@ SEASTAR_TEST_CASE(test_try_catch) {
+        auto inner = std::make_exception_ptr(nested_type("inner"));
+        auto outer1 = make_nested_exception_ptr(nesting_type("outer"), std::move(inner));
+        auto outer2 = make_nested_exception_ptr(utils::internal::default_nested_exception_type("outer2"), std::move(outer1));
+        auto outer3 = make_nested_exception_ptr(utils::internal::default_nested_exception_type("outer2"), std::move(outer2));
```
shouldn't it be "outer3" string? (although it doesn't seem to matter too much here)

```suggestion
        auto outer3 = make_nested_exception_ptr(utils::internal::default_nested_exception_type("outer3"), std::move(outer2));
```

emdroid commented on test/boost/exceptions_test.inc.cc:
```diff
@@ -136,6 +136,73 @@ SEASTAR_TEST_CASE(test_try_catch) {
+    // Negative case, wrong type
+    {
+        auto inner = std::make_exception_ptr(std::string("inner"));
+        auto outer1 = make_nested_exception_ptr(nesting_type("outer"), std::move(inner));
```
could be just `outer` as we only have one here (to be consistent with the "1-level" case), also in the other negative case

gleb-cloudius commented on utils/exceptions.hh:
```diff
@@ -149,6 +151,29 @@ inline T* try_catch(std::exception_ptr& eptr) noexcept {
+        return non_nested;
+    }
+
+    /// Go through all the wrapped levels until we have not found requested exception.
```
Do you mean: "until we find requested exception"?

bitpathfinder replied:
Yep, that's what I meant.

xemul commented:
> Agree, honestly I don't like the solution, since the root cause is the services shutdown order.

@bitpathfinder , I must have missed that comment, but -- did you find out which _exact_ shutdown mis-order causes that to happen?

bitpathfinder commented:
> > Agree, honestly I don't like the solution, since the root cause is the services shutdown order.
> 
> @bitpathfinder , I must have missed that comment, but -- did you find out which _exact_ shutdown mis-order causes that to happen?

Nope, investigating and fixing shutdown sequence would take much more time, meanwhile we have a lot of CI issues which disrupt our dev work. In fact this patch just fixes already existing workaround logic that to ignore these exceptions.

bitpathfinder commented:
> OK, it means that we still don't know if it's shutdown _misorder_ or not. Thanks.

Well imho eaten up exceptions is a clear sign of a _misorder_. Question is, whether it is worth fixing them.

xemul commented:
> Well imho eaten up exceptions is a clear sign of a misorder. Question is, whether it is worth fixing them.

Not necessarily. Proxy may have stopped in order, but "forgot" to clean/join background handlers after itself.

gleb-cloudius commented:
> > Well imho eaten up exceptions is a clear sign of a misorder. Question is, whether it is worth fixing them.
> 
> Not necessarily. Proxy may have ~stopped~ been drained in order, but "forgot" to clean/join background handlers after itself.

Currently sp is not stopped at all. @bhalevy tried to. Not sure what happened to the attempt.

xemul commented:
> Currently sp is not stopped at all.

FYI: not "at all", its "remote" part is stopped and freed on stop (it's not necessarily related to the problem this PR tries to fix)

gleb-cloudius commented:
> > Currently sp is not stopped at all.
> 
> FYI: not "at all", its "remote" part is stopped and freed on stop (it's not necessarily related to the problem this PR tries to fix)

Stopping remote though does not involve waiting for requests to complete. Better than nothing I guess.

bitpathfinder commented:
> * [test_data_distribution_balance[6-IncrementalCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16361/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_consistent_topology_changes___test_data_distribution_balance_6_IncrementalCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance%5B6-IncrementalCompactionStrategy%5D)
> 
> * [test_replace.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16361/testReport/junit/cluster/test_tablets_removenode/Tests___Unit_Tests___test_replace_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace.debug.3)

known flaky tests:  https://github.com/scylladb/scylladb/issues/22745, https://github.com/scylladb/scylladb/issues/20492

bitpathfinder commented:
Restarted CI https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16406/

piodul commented on utils/exceptions.hh:
```diff
@@ -149,6 +151,29 @@ inline T* try_catch(std::exception_ptr& eptr) noexcept {
 
+/// The same as try_catch, but also unwraps if the exception was nested.
+template <typename TException>
+inline TException* try_catch_nested(std::exception_ptr& eptr) noexcept {
```
I am not confident that this function is a good idea in general.

If you have a situation like that:

```
} catch (const Foo&) {
    std::throw_with_nested<Bar>(Bar());
}
```

then the author might have intended that the exception can _only_ be caught as `Bar` and not `Foo`. In particular, I do see a couple of places in the code which do

```
} catch (...) {
    std::throw_with_nested<Bar>(Bar());
}
```

and now it is completely unclear.

I am just worried that `try_catch_nested` might catch too many exceptions that it should, and that it gives some control away from the places that `std::throw_with_nested`.

piodul replied:
That said, I was a bit surprised/disappointed to find that we already use this pattern in some places in our codebase, so I guess the precedent is already set... I won't block merging this PR but I think we really need some clear, written-down guidelines for designing and handling exceptions (maybe we do and people longer in this project already adhere to them, but I just don't get them).

bitpathfinder replied:
IMHO the problem with the nested exceptions - is that it gives too much flexibility, and makes the processing of such exceptions kind of awkward.
I also was tempted to do the trick like that, so at the places where that exception is processed, no changes would be needed. 
```C++
  auto eptr = get_current_exception(); //Should be Bar
  std::throw_with_nested<Bar>(eptr);
```
However for example `gate_closed_exception` does not have any space (string) to put some additional info. So I have to either create a new exception type that inherits gate closed or wrap it in another exception type (eg `runtime_error`) that has additional members to save the info.

piodul commented:
Queued

bitpathfinder commented on utils/exceptions.hh:
```diff
@@ -89,6 +89,8 @@ namespace utils::internal {
 #if defined(OPTIMIZED_EXCEPTION_HANDLING_AVAILABLE)
 void* try_catch_dynamic(std::exception_ptr& eptr, const std::type_info* catch_type) noexcept;
 
+using default_nested_exception_type = std::runtime_error;
```
wrap_commitlog_add_error has the default `std::runtime_error` type and in fact all the exceptions in `replica/database.cc` are wrapped with that type. The idea behind this typedef is to have the `default` type defined explicitly.

bitpathfinder replied:
> wrap_commitlog_add_error has the default `std::runtime_error` type and in fact all the exceptions in `replica/database.cc` are wrapped with that type. The idea behind this typedef is to have the `default` type defined explicitly.

I meant except the 'special' processing for timeout errors.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16406/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16406/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4355/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16406/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7514/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15568/)
#### Build Details:
- Duration: 12 hr
- Builder: spider8.cloudius-systems.com
